### ISSUE TITLE
Figure out how to track base denomination per-asset

### ISSUE THREAD

**Issue by redshiftzero** [penumbra-zone/penumbra#178]
> We should figure out how to keep track of what the base denomination of each asset is, since we'll want to be recording integral values of `upen` or whatever, but displaying as `9.23238921` or something.

_Originally posted by @hdevalence in https://github.com/penumbra-zone/penumbra/issues/176#issuecomment-976016313_

**Comment by redshiftzero** [comment]
Here's thoughts so far: does this seem reasonable/make sense?

## Current situation
- We create assets at genesis.
- There is not currently a way to create new assets. While we do have logic in `pd`'s `State::commit_block` to add new assets to the node asset registry, only assets that have already been created should actually appear here. 

## Proposal
- The current genesis data is put under a `notes` key. 
- We add to the genesis data an `assets` key which follows the same format as the `assets` key in Osmosis' AssetList (https://github.com/osmosis-labs/assetlists). This would mean we would specify for each asset:
            - Description as a string,
            - Denomination units, e.g. for penumbra we can specify the main token penumbra and the minimum denomination (ref #57),
            - What the base denomination should be, e.g. for penumbra if we go for upenumbra (10^{-6}) as the base then we would set the base token as upenumbra and perform all calculations, e.g. value balance, using the base token of the asset,
            - The short symbol, e.g. for penumbra if we did want a PEN short symbol this would be a place to specify it,
            - Optionally links to logo URIs,
- Clients would fetch the asset list from the node registry as usual which now has this additional data, and we'd implement logic on the client-side to do conversions between the denomination units such that transactions are created using the base denomination, but users can specify whichever is convenient in the UI/CLI.

## Later
- In MVP2: at genesis we'd create an asset type for the delegation token `dPEN`. Users will perform `Delegation` and `Undelegation` actions in order to go between `PEN` and `dPEN`.
- In MVP3, a new asset type will be added when an IBC transfer into Penumbra occurs. 
- In MVP4, we will allow assets to be added to the list using the `OpenPosition` (and correspondingly be removed during a `ClosePosition`).
- By scanning the chain from genesis, a new node could reconstruct the asset list from the assets in the genesis file and on-chain activity (IBC transfers in, and positions that have been opened).

**Comment by kwf** [comment]
This mostly matches my understanding! Can we elaborate a bit more about how the different `dPEN` tokens will be represented here? Since each validator's `dPEN` is actually a separate asset, subscripted by that validator's unique identity.

**Comment by redshiftzero** [comment]
Hmm this is a good point. So the first `Delegation` for a new validator will need to create that validators `dPEN` token in the asset registry (ref #37). Then, a new node scanning through the chain will be able to reconstruct the full asset registry from the on-chain IBC transfers, open positions, and delegations, plus the genesis data.

**Comment by avahowell** [comment]
I think the Osmosis AssetList struct already does this, but it might be good to align the design with the Cosmos SDK proposal that implements similar functionality, ADR-024 (https://docs.cosmos.network/master/architecture/adr-024-coin-metadata.html), in case this makes interoperating with Cosmos SDK chains easier in the future.

**Comment by redshiftzero** [comment]
That's a good point, perhaps at first for simplicity we should just use the ADR24 format for the asset metadata (since the osmosis AssetList has additional fields we don't _really_ need e.g. the logos, if we decided we need them perhaps we could add in a future testnet): 

```proto
message DenomUnit {
  string denom    = 1;
  uint32 exponent = 2;  
  repeated string aliases = 3;
}

message Metadata {
  string description = 1;
  repeated DenomUnit denom_units = 2;
  string base = 3;
  string display = 4;
}

```

**Comment by hdevalence** [comment]
There's work on this in multiple directions in both #241 and #242, and I think that having mapped out the space somewhat, it would be good to pause for a bit (cc @zbuc  @redshiftzero) until we have a more cohesive design.

The current code uses `Denom` as an alias for `String` basically as a placeholder, and the new code extends on top of this.  But this causes new issues -- there's new possibility for misuse (what happens if I compute the `asset::Id` corresponding to a non-base `Denom`?), there aren't clear relationships between different denominations, the strings can be modified directly, etc. ADR-024 is helpful, but isn't sufficient for our needs, since (as in #242, #243), we'll need to be able to represent families of assets.

Instead of figuring out how to add on to the existing modeling, I think it would be useful to start mapping out the concepts we need from scratch, and seeing how they fit together.  Here's one sketch:

- `asset::Id`: a (SNARK-friendly) hash of a denomination trace, used in all of the on-chain data structures.  There should only be one of these for each asset.
- `asset::BaseDenom`: the base denomination of an asset. There should only be one of these for each asset.
  * `BaseDenom::id(&self) -> asset::Id` computes an asset ID from the base denomination.
    - By making `Id`s computed only from `BaseDenom`s, we reduce the problem of ensuring there's only one asset ID for each asset to ensuring there's only one `BaseDenom` for each asset.
    - The `asset::Id` could actually be precomputed.
- `asset::DisplayDenom`: a display denomination of an asset.
  * This is kind of like the `DenomUnit` in #241
  * `DisplayDenom::base(&self) -> BaseDenom` could return the associated base denomination, if the display denom stored a copy of the base denom internally.
  * We could also provide a `BaseDenom::units() -> &[DisplayDenom]`, if we committed to having all of the display denominations for a particular asset be known whenever we construct the base denomination.

This (roughly) handles the data that's used in general-purpose asset handling. 

We also need to be able to model families of assets, as in #242. A family of assets could be represented with a type modeling an instance of the family, with a method that returns a `BaseDenom` for that asset.  Since that type represents a specific family, it probably makes sense not to keep it in the `penumbra_crypto::asset` module.

For validator-specific assets, we could have:

- a `DelegationToken` somewhere in the `pd` crate, say `pd::stake::DelegationToken` (for now, I think it could make sense to do a workspace-wide crate reorganization pass soon)
  * Like #242, this would store the validator's identity key internally, and could provide a method for getting the validator the delegation token is associated to.
  * The form of the base denomination is something like `udpenumbra-penumbravaloper1grgelyng2v6v3t8z87wu3sxgt9m5s03xfytvz7`
  * The form of the display denomination is something like `dpenumbra-penumbravaloper1grgelyng2v6v3t8z87wu3sxgt9m5s03xfytvz7`
  * `DelegationToken::denom(&self) -> BaseDenom` could return the base denomination for the delegation token
    * If we provided `units()` for `BaseDenom`, this would also mean constructing a templated set of display denominations.
  * `DelegationToken::parse(...) -> Result<Self, Error>` should parse a `DelegationToken`.  This could take a `&str`, or it could be limited to a specific type like `BaseDenom`. (Since there's already parsing happening, I'm not so worried about taking a `&str`, and it means not forcing a user to parse somewhere else first to get a hold of a `BaseDenom`, but if we take a `&str` we should probably provide some way to get a `&str` representation of an asset)

We could also imagine applying this system to model IBC assets, with a type that stored a parsed representation of the denomination trace, though it's probably not worth trying to go too deep into this now.

However, while it's easy to see how we can go "down" the chain of `DelegationToken` -> `BaseDenom` -> `asset::Id`, it's less clear how to go "up" from less-typed to more-typed data, e.g., how to turn strings into denominations (how do we know whether a given string is a base or display denomination?) or look up the denomination for an asset ID.  Moving "upwards" requires the ability to add extra information, and that information has to come from somewhere.

I'd suggest that we maintain this information as an `AssetRegistry`.  (Where should this live? It's useful for both `pd` and `pcli`, perhaps it should go in `penumbra_crypto` for now, although it's not really cryptographic, hence the sense that a reorganization pass could be good). However, unlike our current setup, we can't just have a list of assets, and their denominations etc., because we need to be able to work with families of assets (e.g., denomination tokens, of which there are in principle arbitrarily many).

The `AssetRegistry` could have:

- A list of regular expressions that match base denominations
  * The [`RegexSet`](https://docs.rs/regex/latest/regex/struct.RegexSet.html) type from the `regex` crate is useful here because it can match an entire set of regular expressions simultaneously.
  * We also need information on the denomination that each regex corresponds to.  For single assets, this can be static data, but for families of assets, this data needs to somehow be parameterized by the (named) regex matches.
  * This is used to provide an `AssetRegistry::parse_base(&self, denom: &str) -> Option<BaseDenom>`
- A list of regular expressions that match display denominations
  * As with base denominations, we need information on the denominations the regexes match, and for asset families this data also needs to be parameterized / templated.
  * This is used to provide an `AssetRegistry::parse_display(&self, denom: &str) -> Option<DisplayDenom>`

This provides a way to pass from strings to parsed and well-typed denominations. Because each family has a single entry, the data required to store the asset registry is basically fixed, until we add additional families of assets. However, we also need a way to go from `asset::Id`s to base denomination strings, and this can't be done using fixed lookup tables.  For instance, if someone sends us some delegation tokens for a particular validator, we just get the asset ID, and if the validator is new, it might not be in a table we currently have, so before we can match it against the registry, we need to get the base denomination string.

An easy way to look up known base denomination strings is `BTreeMap<asset::Id, String>` or similar. For unknown strings, we could use the thin wallet API, although this is a privacy leak.  Another option would be to download the entire collection of base denomination strings to asset IDs (we do this now), though this may not be scalable long-term. This is probably the best thing to do for now, because it is (a) easy and (b) does not require making network requests to figure out what an asset ID represents.

One question is whether the asset registry should do both the ID-to-String lookup and the String-to-Denom parsing.  The advantage is that the API is potentially simpler if there's just one registry that does everything. The disadvantage is that the string-to-denom data (e.g., the regexes, base/display denomination lists) doesn't really need updating, while the id-to-string lookup data does.  If we decided to split them, an `AssetCache` with a `BTreeMap<asset::Id, String>` internally and `query(&self, id: asset::Id) -> Option<&str>`, `update(&mut self, id: asset::Id, denom: String)` methods externally would provided the necessary API; if we decided to unify them, those methods could be on the asset registry.

**Comment by redshiftzero** [comment]
implemented in #283
